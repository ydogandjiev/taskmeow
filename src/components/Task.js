import React, { useRef } from "react";
import { Checkbox, Icon, Link } from "office-ui-fabric-react";
import { useDrag, useDrop } from "react-dnd";
import { ItemTypes } from "./ItemTypes";

const Task = (props) => {
  const ref = useRef(null);
  const [, drop] = useDrop({
    accept: ItemTypes.CARD,
    hover(item, monitor) {
      if (!ref.current) {
        return;
      }
      const dragIndex = item.index;
      const hoverIndex = props.index;

      // Don't replace items with themselves
      if (dragIndex === hoverIndex) {
        return;
      }

      // Determine rectangle on screen
      const hoverBoundingRect = ref.current?.getBoundingClientRect();

      // Get vertical middle

      const hoverMiddleY =
        (hoverBoundingRect.bottom - hoverBoundingRect.top) / 2;

      // Determine mouse position
      const clientOffset = monitor.getClientOffset();

      // Get pixels to the top
      const hoverClientY = clientOffset.y - hoverBoundingRect.top;

      // Only perform the move when the mouse has crossed half of the items height
      // When dragging downwards, only move when the cursor is below 50%
      // When dragging upwards, only move when the cursor is above 50%
      // Dragging downwards
      if (dragIndex < hoverIndex && hoverClientY < hoverMiddleY) {
        return;
      }

      // Dragging upwards
      if (dragIndex > hoverIndex && hoverClientY > hoverMiddleY) {
        return;
      }

      // Time to actually perform the action
      props.onMoveTask(dragIndex, hoverIndex);

      // Note: we're mutating the monitor item here!
      // Generally it's better to avoid mutations,
      // but it's good here for the sake of performance
      // to avoid expensive index searches.
      item.index = hoverIndex;
    },
  });

  const [{ isDragging }, drag] = useDrag({
    item: { type: ItemTypes.CARD, id: props.task._id, index: props.index },
    collect: (monitor) => ({
      isDragging: !!monitor.isDragging(),
    }),
  });

  drag(drop(ref));

  return (
    <div ref={ref} style={{ opacity: isDragging ? 0.5 : 1 }}>
      <li className="Task-listitem">
        <Checkbox
          className="Task-checkbox"
          onChange={(event, isChecked) =>
            props.onCheckedChange(props.task, isChecked)
          }
        />
        <span className="Task-title">
          <div>{props.task.title}</div>
        </span>
        <Link className="Task-star-link">
          {props.task.starred ? (
            <Icon
              className="Task-starred-icon"
              iconName="FavoriteStarFill"
              onClick={(event) => props.onStarredChange(props.task, false)}
            />
          ) : (
            <Icon
              className="Task-unstarred-icon"
              iconName="FavoriteStar"
              onClick={(event) => props.onStarredChange(props.task, true)}
            />
          )}
        </Link>
        {props.inTeams && props.supportsConversation && (
          <Link className="Task-conversation-link">
            {props.task.conversationOpen ? (
              <Icon
                className="Task-conversation-open-icon"
                iconName="ChatSolid"
                onClick={(event) => props.closeConversation(props.task)}
              />
            ) : (
              <Icon
                className="Task-conversation-closed-icon"
                iconName="Chat"
                onClick={(event) => props.openConversation(props.task)}
              />
            )}
          </Link>
        )}
      </li>
    </div>
  );

  /*
  return (
    <li
      className="Task-listitem"
      ref={drag}
      style={{ opacity: isDragging ? 0.5 : 1 }}
    >
      <Checkbox
        className="Task-checkbox"
        onChange={(event, isChecked) =>
          props.onCheckedChange(props.task, isChecked)
        }
      />
      <span className="Task-title">
        <div>{props.task.title}</div>
      </span>
      <Link className="Task-star-link">
        {props.task.starred ? (
          <Icon
            className="Task-starred-icon"
            iconName="FavoriteStarFill"
            onClick={(event) => props.onStarredChange(props.task, false)}
          />
        ) : (
          <Icon
            className="Task-unstarred-icon"
            iconName="FavoriteStar"
            onClick={(event) => props.onStarredChange(props.task, true)}
          />
        )}
      </Link>
      {props.inTeams && props.supportsConversation && (
        <Link className="Task-conversation-link">
          {props.task.conversationOpen ? (
            <Icon
              className="Task-conversation-open-icon"
              iconName="ChatSolid"
              onClick={(event) => props.closeConversation(props.task)}
            />
          ) : (
            <Icon
              className="Task-conversation-closed-icon"
              iconName="Chat"
              onClick={(event) => props.openConversation(props.task)}
            />
          )}
        </Link>
      )}
    </li>
  );
  */
};

export default Task;
