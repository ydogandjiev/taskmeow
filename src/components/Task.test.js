import React from "react";
import { render } from "@testing-library/react";
import { wrapInTestContext } from "react-dnd-test-utils";
import Task from "./Task";

// Initialize Office Fabric icons for use throughout app
import { initializeIcons } from "@uifabric/icons";
initializeIcons();

beforeEach(() => {
  jest.resetModules();
});

it("renders starred task snapshot", () => {
  const mockTask = {
    title: "mockTitle",
    order: 100,
    starred: true,
  };

  const handleTaskCheckedChange = jest.fn();
  const handleTaskStarredChange = jest.fn();

  const TestTask = wrapInTestContext(Task);
  const { container, getByText } = render(
    <TestTask
      task={mockTask}
      onCheckedChange={handleTaskCheckedChange}
      onStarredChange={handleTaskStarredChange}
    />
  );

  expect(getByText(mockTask.title)).toBeInTheDocument();
  expect(container.firstChild).toMatchSnapshot();

  /*
  const taskTitle = (
    <span className="Task-title">
      <div>{mockTask.title}</div>
    </span>
  );

  expect(task).toContainReact(taskTitle);

  expect(task).toContainMatchingElement("li.Task-listitem");
  expect(task).toContainMatchingElement(".Task-checkbox > input");
  expect(task).toContainMatchingElement("button.Task-star-link");
  expect(task).toContainMatchingElement("i.Task-starred-icon");
  expect(task).not.toContainMatchingElement("i.Task-unstarred-icon");
  expect(task).not.toContainMatchingElement("button.Task-conversation-link");

  task.find(".Task-checkbox > input").simulate("change");
  expect(handleTaskCheckedChange).toHaveBeenCalledWith(mockTask, true);

  task.find("i.Task-starred-icon").simulate("click");
  expect(handleTaskStarredChange).toHaveBeenCalledWith(mockTask, false);
  */
});

it("renders unstarred task snapshot", () => {
  const mockTask = {
    title: "mockTitle",
    order: 100,
    starred: false,
  };

  const handleTaskCheckedChange = jest.fn();
  const handleTaskStarredChange = jest.fn();

  const TestTask = wrapInTestContext(Task);
  const { container, getByText } = render(
    <TestTask
      task={mockTask}
      onCheckedChange={handleTaskCheckedChange}
      onStarredChange={handleTaskStarredChange}
    />
  );

  expect(getByText(mockTask.title)).toBeInTheDocument();
  expect(container.firstChild).toMatchSnapshot();

  /*
  const taskTitle = (
    <span className="Task-title">
      <div>{mockTask.title}</div>
    </span>
  );
  expect(task).toContainReact(taskTitle);

  expect(task).toContainMatchingElement("li.Task-listitem");
  expect(task).toContainMatchingElement(".Task-checkbox > input");
  expect(task).toContainMatchingElement("button.Task-star-link");
  expect(task).toContainMatchingElement("i.Task-unstarred-icon");
  expect(task).not.toContainMatchingElement("i.Task-starred-icon");
  expect(task).not.toContainMatchingElement("button.Task-conversation-link");

  task.find(".Task-checkbox > input").simulate("change");
  expect(handleTaskCheckedChange).toHaveBeenCalledWith(mockTask, true);

  task.find("i.Task-unstarred-icon").simulate("click");
  expect(handleTaskStarredChange).toHaveBeenCalledWith(mockTask, true);
  */
});

it("matches starred task with conversation closed snapshot in Teams", () => {
  const mockTask = {
    title: "mockTitle",
    order: 100,
    starred: true,
    conversationOpen: false,
  };

  const handleOpenConversation = jest.fn();

  const TestTask = wrapInTestContext(Task);
  const { container, getByText } = render(
    <TestTask
      task={mockTask}
      inTeams={true}
      supportsConversation={true}
      conversationOpen={false}
      openConversation={handleOpenConversation}
    />
  );

  expect(getByText(mockTask.title)).toBeInTheDocument();
  expect(container.firstChild).toMatchSnapshot();

  /*
  const taskTitle = (
    <span className="Task-title">
      <div>{mockTask.title}</div>
    </span>
  );
  expect(task).toContainReact(taskTitle);

  expect(task).toContainMatchingElement("li.Task-listitem");
  expect(task).toContainMatchingElement(".Task-checkbox > input");
  expect(task).toContainMatchingElement("button.Task-star-link");
  expect(task).toContainMatchingElement("i.Task-starred-icon");
  expect(task).toContainMatchingElement("button.Task-conversation-link");
  expect(task).toContainMatchingElement("i.Task-conversation-closed-icon");
  expect(task).not.toContainMatchingElement("i.Task-conversation-open-icon");

  task.find("i.Task-conversation-closed-icon").simulate("click");
  expect(handleOpenConversation).toHaveBeenCalledWith(mockTask);
  */
});

it("matches unstarred task snapshot with conversation open in Teams", () => {
  const mockTask = {
    title: "mockTitle",
    order: 100,
    starred: true,
    conversationOpen: true,
  };

  const handleCloseConversation = jest.fn();

  const TestTask = wrapInTestContext(Task);
  const { container, getByText } = render(
    <TestTask
      task={mockTask}
      inTeams={true}
      supportsConversation={true}
      conversationOpen={true}
      closeConversation={handleCloseConversation}
    />
  );

  expect(getByText(mockTask.title)).toBeInTheDocument();
  expect(container.firstChild).toMatchSnapshot();

  /*
  const taskTitle = (
    <span className="Task-title">
      <div>{mockTask.title}</div>
    </span>
  );
  expect(task).toContainReact(taskTitle);

  expect(task).toContainMatchingElement("li.Task-listitem");
  expect(task).toContainMatchingElement(".Task-checkbox > input");
  expect(task).toContainMatchingElement("button.Task-star-link");
  expect(task).toContainMatchingElement("i.Task-starred-icon");
  expect(task).toContainMatchingElement("button.Task-conversation-link");
  expect(task).toContainMatchingElement("i.Task-conversation-open-icon");
  expect(task).not.toContainMatchingElement("i.Task-conversation-closed-icon");

  task.find("i.Task-conversation-open-icon").simulate("click");
  expect(handleCloseConversation).toHaveBeenCalledWith(mockTask);
  */
});
