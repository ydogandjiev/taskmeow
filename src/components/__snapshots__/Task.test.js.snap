// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches starred task with conversation closed snapshot in Teams 1`] = `
<div
  style="opacity: 1;"
>
  <li
    class="Task-listitem"
  >
    <div
      class="ms-Checkbox is-enabled Task-checkbox root-78"
    >
      <input
        class="input-79"
        id="checkbox-3"
        type="checkbox"
      />
      <label
        class="ms-Checkbox-label label-80"
        for="checkbox-3"
      >
        <div
          class="ms-Checkbox-checkbox checkbox-81"
        >
          <i
            aria-hidden="true"
            class="ms-Checkbox-checkmark checkmark-85"
            data-icon-name="CheckMark"
            role="presentation"
          >
            
          </i>
        </div>
      </label>
    </div>
    <span
      class="Task-title"
    >
      <div>
        mockTitle
      </div>
    </span>
    <button
      class="ms-Link Task-star-link root-86"
      type="button"
    >
      <i
        aria-hidden="true"
        class="Task-starred-icon root-87"
        data-icon-name="FavoriteStarFill"
        role="presentation"
      >
        
      </i>
    </button>
    <button
      class="ms-Link Task-conversation-link root-86"
      type="button"
    >
      <i
        aria-hidden="true"
        class="Task-conversation-closed-icon root-89"
        data-icon-name="Chat"
        role="presentation"
      >
        
      </i>
    </button>
  </li>
</div>
`;

exports[`matches unstarred task snapshot with conversation open in Teams 1`] = `
<div
  style="opacity: 1;"
>
  <li
    class="Task-listitem"
  >
    <div
      class="ms-Checkbox is-enabled Task-checkbox root-78"
    >
      <input
        class="input-79"
        id="checkbox-4"
        type="checkbox"
      />
      <label
        class="ms-Checkbox-label label-80"
        for="checkbox-4"
      >
        <div
          class="ms-Checkbox-checkbox checkbox-81"
        >
          <i
            aria-hidden="true"
            class="ms-Checkbox-checkmark checkmark-85"
            data-icon-name="CheckMark"
            role="presentation"
          >
            
          </i>
        </div>
      </label>
    </div>
    <span
      class="Task-title"
    >
      <div>
        mockTitle
      </div>
    </span>
    <button
      class="ms-Link Task-star-link root-86"
      type="button"
    >
      <i
        aria-hidden="true"
        class="Task-starred-icon root-87"
        data-icon-name="FavoriteStarFill"
        role="presentation"
      >
        
      </i>
    </button>
    <button
      class="ms-Link Task-conversation-link root-86"
      type="button"
    >
      <i
        aria-hidden="true"
        class="Task-conversation-open-icon root-91"
        data-icon-name="ChatSolid"
        role="presentation"
      >
        
      </i>
    </button>
  </li>
</div>
`;

exports[`renders starred task snapshot 1`] = `
<div
  style="opacity: 1;"
>
  <li
    class="Task-listitem"
  >
    <div
      class="ms-Checkbox is-enabled Task-checkbox root-78"
    >
      <input
        class="input-79"
        id="checkbox-1"
        type="checkbox"
      />
      <label
        class="ms-Checkbox-label label-80"
        for="checkbox-1"
      >
        <div
          class="ms-Checkbox-checkbox checkbox-81"
        >
          <i
            aria-hidden="true"
            class="ms-Checkbox-checkmark checkmark-85"
            data-icon-name="CheckMark"
            role="presentation"
          >
            
          </i>
        </div>
      </label>
    </div>
    <span
      class="Task-title"
    >
      <div>
        mockTitle
      </div>
    </span>
    <button
      class="ms-Link Task-star-link root-86"
      type="button"
    >
      <i
        aria-hidden="true"
        class="Task-starred-icon root-87"
        data-icon-name="FavoriteStarFill"
        role="presentation"
      >
        
      </i>
    </button>
  </li>
</div>
`;

exports[`renders unstarred task snapshot 1`] = `
<div
  style="opacity: 1;"
>
  <li
    class="Task-listitem"
  >
    <div
      class="ms-Checkbox is-enabled Task-checkbox root-78"
    >
      <input
        class="input-79"
        id="checkbox-2"
        type="checkbox"
      />
      <label
        class="ms-Checkbox-label label-80"
        for="checkbox-2"
      >
        <div
          class="ms-Checkbox-checkbox checkbox-81"
        >
          <i
            aria-hidden="true"
            class="ms-Checkbox-checkmark checkmark-85"
            data-icon-name="CheckMark"
            role="presentation"
          >
            
          </i>
        </div>
      </label>
    </div>
    <span
      class="Task-title"
    >
      <div>
        mockTitle
      </div>
    </span>
    <button
      class="ms-Link Task-star-link root-86"
      type="button"
    >
      <i
        aria-hidden="true"
        class="Task-unstarred-icon root-87"
        data-icon-name="FavoriteStar"
        role="presentation"
      >
        
      </i>
    </button>
  </li>
</div>
`;
